### PR TITLE
[CL Message Audit] MsgCollectIncentive 

### PR DESCRIPTION
**Pull request by stackman27**
Clickup
https://app.clickup.com/37420681/v/dc/13nzm9-19160/13nzm9-35260

## What is the purpose of the change
This PR introduces minor additions / changes relating to MsgCreateConcentratedPool:

Test additions / refactors
In-line comment additions / changes
Documentation grooming
Replace manual errors for custom error types
Drive by changes

## Brief Changelog
n/a

## Testing and Verifying
checked all functions were tested

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Comment by stackman27** [comment]
Will continue working on the issues discovered in different PR: 
https://github.com/osmosis-labs/osmosis/issues/5056
https://github.com/osmosis-labs/osmosis/issues/5049

File: x/concentrated-liquidity/incentives.go [hunk 1]
```diff
@@ -599,7 +602,12 @@ func (k Keeper) GetUptimeGrowthInsideRange(ctx sdk.Context, poolId uint64, lower
 	}
 }
 
-// GetUptimeGrowthOutsideRange returns the uptime growth outside the given tick range for all supported uptimes
+// GetUptimeGrowthOutsideRange returns the uptime growth outside the given tick range for all supported uptimes.
+// UptimeGrowthOutside tracks the incentive accured by the entire pool. It keeps track of the cumulative amount of incentives collected
+// by a specific pool since the last time incentives were accured.
+// We use this function to calculate to the total amount of incentives owed to the LPs when they withdraw their liquidity or when they
+// attempt to claim their incentives.
+// When LPs are rady to claim their incentives we calculate it using: (shares of # of LP) * (uptimeGrowthOutside - uptimeGrowthInside)
```
Comment (stackman27) [inline x/concentrated-liquidity/incentives.go:610]:
```suggestion
// When LPs are ready to claim their incentives we calculate it using: (shares of # of LP) * (uptimeGrowthOutside - uptimeGrowthInside)
```

File: x/concentrated-liquidity/client/cli/tx.go [hunk 2]
```diff
@@ -82,7 +82,7 @@ func NewCreateIncentiveCmd() (*osmocli.TxCliDesc, *types.MsgCreateIncentive) {
 	return &osmocli.TxCliDesc{
 		Use:                 "create-incentive [incentive-denom] [incentive-amount] [emission-rate] [start-time] [min-uptime]",
 		Short:               "create an incentive record to emit incentives (per second) to a given pool",
-		Example:             "create-incentive uosmo 69082 0.02 100 2023-03-03 03:20:35.419543805 24h --pool-id 1 --from val --chain-id osmosis-1",
+		Example:             "create-incentive uosmo 69082 0.02 2023-03-03 03:20:35.419543805 24h --pool-id 1 --from val --chain-id osmosis-1",
 		CustomFlagOverrides: poolIdFlagOverride,
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/client/cli/tx.go:86]: Can we create an issue to not require pool ID flag over ride and just accept it as an input param?

File: x/concentrated-liquidity/incentives.go [hunk 3]
```diff
@@ -28,6 +27,7 @@ func (k Keeper) createUptimeAccumulators(ctx sdk.Context, poolId uint64) error {
 	return nil
 }
 
+// TODO: write go test
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:30]: Is there an issue for this?

File: x/concentrated-liquidity/incentives.go [hunk 4]
```diff
@@ -599,7 +602,12 @@ func (k Keeper) GetUptimeGrowthInsideRange(ctx sdk.Context, poolId uint64, lower
 	}
 }
 
-// GetUptimeGrowthOutsideRange returns the uptime growth outside the given tick range for all supported uptimes
+// GetUptimeGrowthOutsideRange returns the uptime growth outside the given tick range for all supported uptimes.
+// UptimeGrowthOutside tracks the incentive accured by the entire pool. It keeps track of the cumulative amount of incentives collected
+// by a specific pool since the last time incentives were accured.
+// We use this function to calculate to the total amount of incentives owed to the LPs when they withdraw their liquidity or when they
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:608]:
```suggestion
// We use this function to calculate the total amount of incentives owed to the LPs when they withdraw their liquidity or when they
```

File: x/concentrated-liquidity/incentives.go [hunk 5]
```diff
@@ -767,6 +775,7 @@ func (k Keeper) claimAllIncentivesForPosition(ctx sdk.Context, positionId uint64
 
 			// If the claimed incentives are forfeited, deposit them back into the accumulator to be distributed
 			// to other qualifying positions.
+			// NOTE: we cannot hit this logic with current implementation because position gets fully charged at 1nanosecond.
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/incentives.go:778]: Im not sure if this is totally true. If we enter and leave in the same block, even if it takes seconds, we go off block time which is the same and therefore position age would be less than the supported uptime.

File: x/concentrated-liquidity/msg_server.go [hunk 6]
```diff
@@ -135,7 +135,7 @@ func (server msgServer) CollectFees(goCtx context.Context, msg *types.MsgCollect
 	return &types.MsgCollectFeesResponse{CollectedFees: totalCollectedFees}, nil
 }
 
-// CollectIncentives collects incentives for all positions in given range that belong to sender
+// CollectIncentives collects incentives for all given PositionIds in a given range that belongs to sender.
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/msg_server.go:138]:
```suggestion
// CollectIncentives collects incentives for all given PositionIds in a given range that belong to sender.
```

File: x/concentrated-liquidity/types/errors.go [hunk 7]
```diff
@@ -45,6 +45,23 @@ func (e NotPositiveRequireAmountError) Error() string {
 	return fmt.Sprintf("Required amount should be positive. Got: %s", e.Amount)
 }
 
+type QualifyingLiquidityOrTimeElapsedNotPositiveError struct {
+	QualifyingLiquidity sdk.Dec
+	TimeElapsed         sdk.Dec
+}
+
+func (e QualifyingLiquidityOrTimeElapsedNotPositiveError) Error() string {
+	return fmt.Sprintf("Qualifying liquidity and time elapsed must both be positive. Got: QualifyingLiquidity%s, timeElapsed%s", e.QualifyingLiquidity, e.TimeElapsed)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/errors.go:54]:
```suggestion
	return fmt.Sprintf("Qualifying liquidity and time elapsed must both be positive. Got: QualifyingLiquidity (%s), timeElapsed (%s)", e.QualifyingLiquidity, e.TimeElapsed)
```

File: x/concentrated-liquidity/types/errors.go [hunk 8]
```diff
@@ -45,6 +45,23 @@ func (e NotPositiveRequireAmountError) Error() string {
...
+}
+
+type TimeElapsedNotPositiveError struct {
+	TimeElapsed sdk.Dec
+}
+
+func (e TimeElapsedNotPositiveError) Error() string {
+	return fmt.Sprintf("Time elapsed must both be positive. Got: timeElapsed%s", e.TimeElapsed)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/types/errors.go:62]:
```suggestion
	return fmt.Sprintf("Time elapsed must both be positive. Got: timeElapsed (%s)", e.TimeElapsed)
```

**Review by czarcas7ic — APPROVED** [review]
These changes LGTM 👍

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (stackman27) [reply x/concentrated-liquidity/incentives.go:30]: it's not too much so i think i'll just write the test in this PR

File: x/concentrated-liquidity/client/cli/tx.go (hunk 2 above)
Comment (stackman27) [reply x/concentrated-liquidity/client/cli/tx.go:86]: sure, what's the main reasoning behind it tho? should we change it for `CreatePosition` too?

File: x/concentrated-liquidity/incentives.go (hunk 5 above)
Comment (stackman27) [reply x/concentrated-liquidity/incentives.go:778]: yup removed

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (stackman27) [reply x/concentrated-liquidity/incentives.go:30]: yessir https://github.com/osmosis-labs/osmosis/issues/5088

File: x/concentrated-liquidity/incentives.go (hunk 5 above)
Comment (stackman27) [inline x/concentrated-liquidity/incentives.go:778]:
```suggestion
```

File: x/concentrated-liquidity/client/cli/tx.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/client/cli/tx.go:86]: We want to move away from required flags whenever possible (anything required should just be an input param).

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/incentives.go:30]: Can you link this issue directly in the TODO as well please?

**Review by p0mvn — APPROVED** [review]
LGTM

File: x/concentrated-liquidity/incentives.go (hunk 3 above)
Comment (p0mvn) [reply x/concentrated-liquidity/incentives.go:30]: @stackman27 please ping for merge once added

Comment (stackman27) [reply x/concentrated-liquidity/incentives.go:30]: test already got merged https://github.com/osmosis-labs/osmosis/pull/5090

Comment (stackman27) [inline x/concentrated-liquidity/incentives.go:30]:
```suggestion
```

**Comment by czarcas7ic** [comment]
Merging since we have two approvals!

**Pull request merged**